pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/WhenStageTest.java

```diff
@@ -122,4 +131,116 @@ public void paramsInWhenExpression() throws Exception {
+                .logNotContains("Stage 'Two' skipped due to when conditional", "Warning, empty changelog.")
+                .go();
+    }
+
+    @Test
+    public void whenChangelog() throws Exception {
+        //TODO JENKINS-46086 First time build always skips the changelog
+        final ExpectationsBuilder builder = expect("when/changelog", "changelog")
+                .logContains("Hello", "Stage 'Two' skipped due to when conditional", "Warning, empty changelog. Probably because this is the first build.")
+                .logNotContains("Dull World");
+        builder.go();
+
+        builder.resetForNewRun(Result.SUCCESS);
+
+        sampleRepo.write("something.txt", "//fake file");
+        sampleRepo.git("add", "something.txt");
+        sampleRepo.git("commit", "-m", "Some title that we don't care about\n\nSome explanation\n[DEPENDENCY] some-app#45");
+
+        builder.logContains("Hello", "Dull World")
+                .logNotContains("Stage 'Two' skipped due to when conditional", "Warning, empty changelog.")
+                .go();
+    }
+
+    @Test
+    public void whenChangelogPR() throws Exception {
+        //TODO JENKINS-46086 First time build always skips the changelog
+        final ExpectationsBuilder builder = expect("when/changelog", "changelog")
+                .logContains("Hello", "Stage 'Two' skipped due to when conditional", "Warning, empty changelog. Probably because this is the first build.")
+                .logNotContains("Dull World");
+        builder.go();
+
+        builder.resetForNewRun(Result.SUCCESS);
+
+        sampleRepo.write("something.txt", "//fake file");
+        sampleRepo.git("add", "something.txt");
+        sampleRepo.git("commit", "-m", "Some title that we don't care about\n\nSome explanation\n[DEPENDENCY] some-app#45");
+
+        builder.logContains("Hello", "Dull World")
+                .logNotContains("Stage 'Two' skipped due to when conditional", "Warning, empty changelog.")
+                .go();
+
+        builder.resetForNewRun(Result.SUCCESS);
+        fakePRMarker(builder.getRun().getParent());
+
+        sampleRepo.write("something2.txt", "//fake file");
+        sampleRepo.git("add", "something2.txt");
+        sampleRepo.git("commit", "-m", "Some title");
+
+        builder.logContains("Hello", "Dull World", "Examining changelog from all builds of this change request") //Should go for the log in build 2
+                .logNotContains("Stage 'Two' skipped due to when conditional", "Warning, empty changelog.")
+                .go();
+
+    }
+
+
+    static void fakePRMarker(WorkflowJob job) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException, IOException {
+        //Black magic ahead!!
+        final Constructor<PullRequestSCMHead> pullRequestSCMHeadConstructor = PullRequestSCMHead.class.getDeclaredConstructor(String.class, boolean.class, int.class, BranchSCMHead.class, String.class, String.class, String.class);
```

Comment:
Gah!
I've spent most of the day rewriting the tests to use `MockSCMSource` and then it turns out that it doesn't support `affectedPaths` in the changelog. _sigh_.